redfish-core/lib/power_supply_metrics.hpp

```diff
@@ -0,0 +1,350 @@
+                    messages::internalError(aResp->res);
+                    return;
+                }
+
+                const std::string& validPath = objectPath;
+                auto psuMatchStr = powerSupplyID + "_input_power";
+                std::string psuInputPowerStr;
+                // validPath -> {psu}_input_power
+                // 5 below comes from
+                // /org/open_power/sensors/aggregation/per_30s/
+                //   0      1         2         3         4
+                // .../{psu}_input_power/[average,maximum]
+                //           5..............6
+                if (!dbus::utility::getNthStringFromPath(validPath, 5,
+                                                         psuInputPowerStr))
+                {
+                    BMCWEB_LOG_ERROR << "Got invalid path " << validPath;
+                    messages::invalidObject(aResp->res, validPath);
+                    return;
+                }
+
+                if (psuInputPowerStr != psuMatchStr)
+                {
+                    // not this power supply, continue to next objectPath...
+                    continue;
+                }
+
+                BMCWEB_LOG_DEBUG << "Got validPath: " << validPath;
+                for (const auto& connection : connectionNames)
+                {
+                    serviceName = connection.first;
+
+                    for (const auto& interfaceName : connection.second)
+                    {
+                        if (interfaceName == "org.open_power.Sensor."
+                                             "Aggregation.History.Average")
+                        {
+                            averagePath = validPath;
+                        }
+                        else if (interfaceName == "org.open_power.Sensor."
+                                                  "Aggregation.History.Maximum")
+                        {
+                            maximumPath = validPath;
+                        }
+                    }
+                }
+            }
+
+            getAverageMaximumValues(aResp, serviceName, averagePath,
+                                    maximumPath);
+        },
+        "xyz.openbmc_project.ObjectMapper",
+        "/xyz/openbmc_project/object_mapper",
+        "xyz.openbmc_project.ObjectMapper", "GetSubTree",
+        "/org/open_power/sensors/aggregation/per_30s", 0, interfaces);
+}
+
+/**
+ * Systems derived class for delivering OemPowerSupplyMetrics Schema.
+ */
+inline void requestRoutesPowerSupplyMetrics(App& app)
+{
+    BMCWEB_ROUTE(app, "/redfish/v1/Chassis/<str>/PowerSubsystem/"
+                      "PowerSupplies/<str>/Metrics")
+        .privileges({{"Login"}})
+        .methods(boost::beast::http::verb::get)(
+            [](const crow::Request&,
+               const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+               const std::string& chassisID, const std::string& powerSupplyID) {
```

Comment:
Do you think there is a way to throw a resourceNotFound error for powerSupplyID like you do for chassis on line 319?
I see checking on line 263, but maybe the logic doesn't an easy/efficient way for this check.

Reply:
This sounds like the getValidPowerSupplyID() that Shawn had suggested. There is such a function, but it was in power_supply.hpp (I think), not in utils. Gunnar had suggested I could move it to a new utils file and use it that way.